### PR TITLE
FIX Calculate threshold condition with SQL rather than PHP

### PR DESCRIPTION
**Pull request by ScopeyNZ**
This is a performance fix. Modern SQL engines can avoid counting a whole result set (potentially thousands of records) when you are only interested if the count exceeds a threshold.

Split from #8915. This is now targeting 4.3 as a patch. Happy to retarget any other branch if requested 🙂

### CODE REVIEW

File: src/ORM/FieldType/DBForeignKey.php [hunk 1]
```diff
@@ -87,12 +87,29 @@ public function scaffoldFormField($title = null, $params = null)
...
+            $distinct = $dataQuery->getDistinct();
+            $count = '*';
+
+            if ($distinct) {
+                $dataQuery->setDistinct(false);
+                $count = 'DISTINCT *';
+            }
+
+            $dataQuery->setOrderBy(false);
+            $dataQuery->setSelect(['over_threshold' => 'count('. $count . ') > ' . (int) $threshold]);
```
Comment (robbieaverill) [inline src/ORM/FieldType/DBForeignKey.php]: Potentially opinionated, but if `count()` is a SQL count function, do you think it's worth writing it uppercase?

File: src/ORM/FieldType/DBForeignKey.php [hunk 2]
```diff
@@ -87,12 +87,29 @@ public function scaffoldFormField($title = null, $params = null)
...
+            $result = $dataQuery->execute()->column('over_threshold')[0];
```
Comment (robbieaverill) [inline src/ORM/FieldType/DBForeignKey.php]: Will the column() call always return a non-empty array?

File: src/ORM/FieldType/DBForeignKey.php [hunk 3]
```diff
@@ -87,12 +87,29 @@ public function scaffoldFormField($title = null, $params = null)
...
             static::$foreignListCache[$hasOneClass] = [
-                'count' => $list->count(),
+                'overThreshold' => is_bool($result) ? $result : ($result === '1' || $result === 't'),
```
Comment (robbieaverill) [inline src/ORM/FieldType/DBForeignKey.php]: Worth adding a note that the `1` `t` checks are for B/C with Postgres < 2.2?

Comment (ScopeyNZ) [reply src/ORM/FieldType/DBForeignKey.php]: Yep I'll do that.

File: src/ORM/FieldType/DBForeignKey.php (hunk 2 above)
Comment (ScopeyNZ) [reply src/ORM/FieldType/DBForeignKey.php]: I think as it's not a subselect that it can. Either way I should be defensively programming.

**Comment by ScopeyNZ** [comment]
Tests are failing because I forgot how to count distinct records. This is really driving me towards addressing #4923 .

As this is a DataObject query it'd be easiest to be like `count(DISTINCT ID)` or something, but I know that you can't assume `ID` is on the table - let alone a distinct column in this instance. I'll think about this a little further I guess.

**Comment by ScopeyNZ** [comment]
Ok this PR is pointless with distinct queries. We can readdress this once #4923 is complete.

**Pull request closed without merging**